### DISCUSSION TITLE
I want to restrict access by IP address.

### DISCUSSION THREAD

**Discussion by ko-morimoto** [keycloak/keycloak#22016] [Q&A]
Hello. 

I would like to achieve access restriction by IP address on keycloak. From my research, it seems that this can be achieved with SPI extensions using Java, but I am not sure.

Please give me some advice.
Thank you in advance.

**Comment by MagiiEl** [comment]
Yes you will have to create an SPI (or find one that works for you). Depending on your experience with programming and Keycloak, it is actually quite simple to implement. And if you have zero experience it is a good introduction to SPI. 

I cannot give you what I have implemented but there is one youtube video covering your subject: https://www.youtube.com/watch?v=u36QK9oyrtM&t=39s . It is probably a little outdated. You shoud refer to the latest documentations for SPIs and Java Docs for Keycloak.

For the Docs, checkout the Server Developer Guide, 8) Authentication SPI, https://www.keycloak.org/docs/latest/server_development/#_auth_spi  . It's best if you read the complete Chapter but essentially you want the part about "Authenticator SPI walk through"

In the Java Docs you should checkout where the HTTP Headers are saved and how to access them. You can probably research how to  handle IP Adresses  / masks formats in java. It's probably up to your taste how to handle that.

My project essentially consists of the META-INF file for the Authenticator Factory. An implemented "Authenticator Factory" Class and the "Authenticator" Class. That's basically it. ( I dont know how outdated that structure for newer keycloaks is though). I also added some message files for error messages in different languages, that's it.

If you don't know already, you should also research how secure an IP address restriction is, when it actually makes sense and when it is easily overcome. E.g. you should never accept an IP Address directly given by the user as it is easily rewritten. 

Hope this will give you an easy start. :) Maybe you are lucky and you find some newer SPI online. But definitely read and checkout any code you find instead of blindly using it.

**Comment by darius-m** [comment]
It depends on what you wish to achieve with the restrictions. If you simply need to allow or deny requests based on the source IP address, it may be easier to use a reverse proxy and use their allow / accept and block / deny rules. However, you may also need to add other configurations (make Keycloak aware that it's behind a reverse proxy, override `X-Forwarded-*` and `Forwarded` headers on the reverse proxy) if you decide to use this approach.

**Comment by danielFesenmeyer** [comment]
I suggest you have a look at this PR (which did not make it into Keycloak): https://github.com/keycloak/keycloak/pull/16497
Filtering the requests on quarkus level works quite fine for us.

There are is also another interesting discussion on this: https://github.com/keycloak/keycloak/discussions/12481

**Comment by hariubc** [comment]
If you have Apache and/or load balancer fronting Keycloak, you can block there.

**Comment by ko-morimoto** [comment]
Dear all

Hello.
Thank you for all the messages.
I will check them carefully one by one.

I would like to control access permission and denial by source IP address. I think one way is to use a reverse proxy as you commented.

If possible, I would like to realize access restriction by IP address for each client. For example, we would like to allow access to service A and deny access to service B, even though they have the same IP address.

First, I will read your comments and consider which means would be best.
As a result, I may consult with you again.

Many thanks for the follow-up, I sincerely appreciate it.
Thank you.
